POC: multiple model/configuration DeepSpeed support (#3097)
huggingface/accelerate: docs/source/usage_guides/deepspeed_multiple_model.md

Reviewer: Very nice examples, thanks for putting in the effort. I leave it to others to comment on the concrete examples, as I didn't have those use cases yet.

A small comment: There are some line breaks here and there that are not necessary, e.g. lines 35-36. But no big deal.

the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contains specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.
-->

# Using multiple models with DeepSpeed

<Tip warning={true}>

    This guide assumes that you have read and understand the [DeepSpeed usage guide](./deepspeed.md).

</Tip>

## Use cases

There are a few use cases where one may want to run multiple models when using `accelerate` and `DeepSpeed`:

* Knowledge distillation
* RLHF techniques (see the `trl` repo for [more examples](https://github.com/huggingface/trl))
* Training multiple models at once.

As it stands, Accelerate has a **very experimental API** to help you accomplish this. 

This tutorial will focus on two common use cases:

> 1. Using `zero2` to train a model, and `zero3` to perform inference for assisting of the training for the `zero2` model (as `zero2` is faster to train on)

Reviewer: Mention "RLHF" to make it clear that this example refers to the next section.

2. Training multiple *disjoint* models at once

## RLHF (multiple models, only one is trained)

Normally, you would use a single [`utils.DeepSpeedPlugin`] at a time, however in this case we have two seperate configurations. Accelerate allows you to create and use multiple plugins **if and only if** they are in a `dict` so that we can reference/enable the proper plugin when needed:

```python
from accelerate.utils import DeepSpeedPlugin

zero2_plugin = DeepSpeedPlugin(hf_ds_config="zero2_config.json")
zero3_plugin = DeepSpeedPlugin(hf_ds_config="zero3_config.json")

> deepspeed_plugins = {"student": zero2_plugin, "teacher": zero3_plugin, }

Reviewer: ```suggestion
deepspeed_plugins = {"student": zero2_plugin, "teacher": zero3_plugin}
```

```

> When doing so,  `zero2_config.json` should be configured for full training (so specifying `scheduler` and `optimizer` if not utilizing your own) while `zero3_config.json` should *only* be configured for the model,

Reviewer: ```suggestion
When doing so, `zero2_config.json` should be configured for full training (so specifying `scheduler` and `optimizer` if not utilizing your own) while `zero3_config.json` should *only* be configured for the model,
```

        "stage": 3,
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "overlap_comm": true,
        "sub_group_size": 1e9,
        "reduce_bucket_size": 1e9,
        "stage3_prefetch_bucket_size": 1e9,
        "stage3_param_persistence_threshold": 1e9,
        "stage3_max_live_parameters": 1e9,
        "stage3_max_reuse_distance": 1e9,
        "stage3_gather_16bit_weights_on_model_save": true
    },
    "train_micro_batch_size_per_gpu": 1
}
```

<Tip>

    DeepSpeed will raise an error if we don't specify a `train_micro_batch_size_per_gpu`, despite not really 
    training this particular model

</Tip>

From here, we can create a single [`Accelerator`] and pass in both configurations:

```python
from accelerate import Accelerator

accelerator = Accelerator(deepspeed_plugin=deepspeed_plugins)
```

Now let's see how to use them.

### The student model

By default, `accelerate` will set the first item in the `dict` as the default/enabled plugin (so here our `"student"` plugin). We can verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin]` function to see which one is being used:

```python

> active_plugin = get_active_deepspeed_plugin(accelerator.state)

Reviewer: Can we not use `accelerator.state.deepspeed_plugin` or `accelerator.deepspeed_plugin`. Or is the idea to make it more explicit?

Author: Idea was to make it more explicit. (I'll add a comment about `accelerator.deepspeed_plugin` though)

assert active_plugin is deepspeed_plugins["student"]
```

Since the `student` is the currently active plugin, let's go ahead and prepare our model, optimizer, and scheduler like normal:

```python
student_model, optimizer, scheduler = ...
student_model, optimizer, scheduler, train_dataloader = accelerator.prepare(student_model, optimizer, scheduler, train_dataloader)
```

From here, we need to deal with the teacher model.

### The teacher model

First we need to tell the `Accelerator` that the zero3/teacher ZeRO configuration should be used:

```python
accelerator.state.enable_deepspeed_plugin("teacher")
```

Doing so will disable the `"student"` plugin and enable the `"teacher"` one instead. This will update the

> DeepSpeed stateful config inside of `transformers`, and change which plugin configuration gets called when using

Reviewer: It's not clear how `transformers` comes into play here. What if the user is not using transformers? It should still work, right?

Author: Yes and no. If not using `transformers` they need to manually call `ZeroInit`, which is what this is talking about

```python
teacher_model.eval()
student_model.train()
for batch in train_dataloader:
    with torch.no_grad():
        output_teacher = teacher_model(**batch)
    output_student = student_model(**batch)
    # Combine the losses or modify it in some way
    loss = output_teacher.loss + output_student.loss
    accelerator.backward(loss)
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
```

## Multiple models all being trained

Multiple models being trained is a more complicated scenario and is actively being looked upon by the team.
In its current state, we assume that each model **completely disjoint** from the other during training.

With this situation, we still require two [`utils.DeepSpeedPlugin`]'s to be made, however we also further
require a second `Accelerator`, since we need to call the proper `deepspeed` engine, and `Accelerator` only
carries one instance of it at a time.

Since the [`state.AcceleratorState`] is a stateful object though, it is already aware of both `DeepSpeedPlugin`'s available,
meaning you can just instantiate a second `Accelerator` with no extra arguments:

```python
accelerator_0 = Accelerator(deepspeed_plugin=deepspeed_plugins)

> accelerator_1 = Accelerator()

Author: Well, now we'll raise an error :)

specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contains specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.
-->

# Using multiple models with DeepSpeed

<Tip warning={true}>

    This guide assumes that you have read and understand the [DeepSpeed usage guide](./deepspeed.md).

</Tip>

## Use cases

There are a few use cases where one may want to run multiple models when using `accelerate` and `DeepSpeed`:

* Knowledge distillation
* RLHF techniques (see the `trl` repo for [more examples](https://github.com/huggingface/trl))
* Training multiple models at once.

As it stands, Accelerate has a **very experimental API** to help you accomplish this.

This tutorial will focus on two common use cases:

1. Using `zero2` to train a model, and `zero3` to perform inference for assisting of the training for the `zero2` model (as `zero2` is faster to train on), which is shown in the RLHF example
2. Training multiple *disjoint* models at once, shown later

> ## RLHF (multiple models, only one is trained)

Reviewer 2: ```suggestion
## Multiple models, only one is trained

RLHF is a good example of such use-case
```

The documentation is about many-models-one-trained, not RLHF - so proposing ot re-arrange this

deepspeed_plugins = {"student": zero2_plugin, "teacher": zero3_plugin}
```

When doing so, `zero2_config.json` should be configured for full training (so specifying `scheduler` and `optimizer` if not utilizing your own) while `zero3_config.json` should *only* be configured for the model,
such as:

```json
{
    "bf16": {
        "enabled": "auto"
    },
    "zero_optimization": {
        "stage": 3,
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "overlap_comm": true,
        "sub_group_size": 1e9,
        "reduce_bucket_size": 1e9,
        "stage3_prefetch_bucket_size": 1e9,
        "stage3_param_persistence_threshold": 1e9,
        "stage3_max_live_parameters": 1e9,
        "stage3_max_reuse_distance": 1e9,
        "stage3_gather_16bit_weights_on_model_save": true
    },
    "train_micro_batch_size_per_gpu": 1
}
```

<Tip>

    DeepSpeed will raise an error if we don't specify a `train_micro_batch_size_per_gpu`, despite not really 
    training this particular model

</Tip>

From here, we can create a single [`Accelerator`] and pass in both configurations:

```python
from accelerate import Accelerator

> accelerator = Accelerator(deepspeed_plugin=deepspeed_plugins)

Reviewer 2: can this be fixed to `deepspeed_plugins` as the key, otherwise this is a poor API.

Perhaps for BC it could be either?

Reviewer 3: This is a DS bug due to sanity checking never getting updated for inference. Please share a repro and issue. Thanks!

```

Now let's see how to use them.

### The student model

By default, `accelerate` will set the first item in the `dict` as the default/enabled plugin (so here our `"student"` plugin). We can verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin]` function to see which one is being used:

```python
active_plugin = get_active_deepspeed_plugin(accelerator.state)
assert active_plugin is deepspeed_plugins["student"]
```

Also the `AcceleratorState` will keep the active DeepSpeed plugin saved in `state.deepspeed_plugin`:
```python
assert active_plugin is accelerator.deepspeed_plugin
```

Since the `student` is the currently active plugin, let's go ahead and prepare our model, optimizer, and scheduler like normal:

```python
student_model, optimizer, scheduler = ...
student_model, optimizer, scheduler, train_dataloader = accelerator.prepare(student_model, optimizer, scheduler, train_dataloader)
```

From here, we need to deal with the teacher model.

### The teacher model

First we need to tell the `Accelerator` that the zero3/teacher ZeRO configuration should be used:

```python

> accelerator.state.enable_deepspeed_plugin("teacher")

Reviewer 2: yeah, now reading the demo, I think I agree with the earlier discussion - `enable` is a bad terminology for something that requires only 1 of many state. Enable implies that more than one thing can be enabled at the same time, but this is not the case.

Toggle could work for 2 objects. But then we may have more than 2 models, in which case toggle doesn't fit either.

So probably `select` is the right verb here, as it implies one among many.

But yet you did say in the docs that only one plugin can be enabled at any given time... not sure. I just think select is more self-documenting.

Author: Will move forward with `.select()` then. If we get to a point where for some reason we have multiple active, we can deprecate and change it back (or just add)

```

Doing so will disable the `"student"` plugin and enable the `"teacher"` one instead. This will update the
DeepSpeed stateful config inside of `transformers`, and change which plugin configuration gets called when using

> `deepspeed.initialize()`, allowing us to use the no-code `Zero3Init` `deepspeed` context manager `transformers` provides:

Reviewer 2: ```suggestion
`deepspeed.initialize()`, allowing us to use the no-code `deepspeed.zero.Init`  context manager `transformers` provides:
```

Reviewer 2: Also, I don't know what "no-code" means

```python
teacher_model = AutoModel.from_pretrained(...)
teacher_model = accelerator.prepare(teacher_model)
```

Otherwise you should manually initialize them under `deepspeed.zero.Init`:
```python
with deepspeed.zero.Init(accelerator.deepspeed_plugin.config):
    model = MyModel(...)
```

### Training

From here, your training loop can be whatever you like, making sure that `teacher_model` is never being trained on:

```python
teacher_model.eval()
student_model.train()
for batch in train_dataloader:
    with torch.no_grad():
        output_teacher = teacher_model(**batch)
    output_student = student_model(**batch)
    # Combine the losses or modify it in some way
    loss = output_teacher.loss + output_student.loss
    accelerator.backward(loss)
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
```

## Multiple models all being trained

Multiple models being trained is a more complicated scenario and is actively being looked upon by the team.

> In its current state, we assume that each model **completely disjoint** from the other during training.

Reviewer 2: ```suggestion
In its current state, we assume that each model is **completely disjoint** from the other during training.
```

perhaps add a note or an example of what you mean by disjoint?

With this situation, we still require two [`utils.DeepSpeedPlugin`]'s to be made, however we also further

> require a second `Accelerator`, since we need to call the proper `deepspeed` engine, and `Accelerator` only

Reviewer 2: ```suggestion
require a second `Accelerator`, since we need to call different `deepspeed` engines at different times, and a single `Accelerator` only
```

carries one instance of it at a time.

Since the [`state.AcceleratorState`] is a stateful object though, it is already aware of both `DeepSpeedPlugin`'s available,
meaning you can just instantiate a second `Accelerator` with no extra arguments:

```python
accelerator_0 = Accelerator(deepspeed_plugin=deepspeed_plugins)
accelerator_1 = Accelerator()
```

Similar to before, we can call either `accelerator_0.state.enable_deepspeed_plugin()` to enable/disable
a particular plugin, and call `.prepare` like normal:

```python
# can be `accelerator_0`, `accelerator_1`, or by calling `AcceleratorState().enable_deepspeed_plugin(...)`

> accelerator_0.state.enable_deepspeed_plugin("model1")

Reviewer 2: if you are using the `_\d` pattern can we use it consistently? i.e. `model_1`, etc? or w/o `_` but then again consistently w/o it.

Reviewer 2: i think the accounting is weird too, you have model0 and model1 and model2 - something is off here - you should have only 2 models here, no?

Author: Fixed :)

model_0 = AutoModel.from_pretrained(...)
# For this example, `get_training_items` is a nonexistent function that gets the setup we need for training
optimizer_0, scheduler_0, train_dl, eval_dl = get_training_items(model_0)
model_0, optimizer_0, scheduler_0, train_dl, eval_dl = accelerator.prepare(model_0, optimizer_0, scheduler_0, train_dl, eval_dl)

accelerator_1.state.enable_deepspeed_plugin("model2")
model_1 = AutoModel.from_pretrained(...)
# For this example, `get_training_items` is a nonexistent function that gets the setup we need for training
optimizer_1, scheduler_1, _, _ = get_training_items(model1)
model_1, optimizer_1, scheduler_1 = accelerator.prepare(model1, optimizer1, scheduler1)
```

And now you can train:

```python
for batch in dl:
    outputs_1 = model_0(**batch)
    accelerator_0.backward(outputs_1.loss)
    optimizer_0.step()
    scheduler_0.step()
    optimizer_0.zero_grad()
    outputs_2 = model_1(**batch)
    accelerator_1.backward(outputs_2.loss)
    optimizer_1.step()
    scheduler_1.step()
    optimizer_1.zero_grad()

> ```

Reviewer 2: I find this example very difficult to follow since the numbers don't match.

perhaps once you explained why this is so and they have to be that way, non-numerical differentiator would work better? perhaps the same teacher/student naming would be more appropriate?

also we are facing humans here, model_0 and model_1 isn't how we think - we think model_1 and model_2 instead. I'd suggest to leave 0-indexing to computer indices.

Reviewer 2: also adding some vertical white space to group things together would help the readability in this last code section.

Author: Went with `first_` and `second_` to be more readable (and the prior example used words)

<Tip warning={true}>

    This guide assumes that you have read and understood the [DeepSpeed usage guide](./deepspeed.md).

</Tip>

Running multiple models with Accelerate and DeepSpeed is useful for:

* Knowledge distillation
* RLHF techniques (see the [TRL](https://github.com/huggingface/trl) library for more examples)
* Training multiple models at once

Currently, Accelerate has a **very experimental API** to help you use multiple models.

This tutorial will focus on two common use cases:

1. An RLHF use case, using ZeRO2 to train a model and ZeRO3 for inference to assist in training the ZeRO2 model. ZeRO2 is faster to train with if the model fits on a single GPU.
2. Training multiple *disjoint* models at once.

## RLHF

RLHF is a good example of using multiple models, but only training one of them.

Normally, you would use a single [`utils.DeepSpeedPlugin`] at a time. However, in this case, there are two separate configurations. Accelerate allows you to create and use multiple plugins **if and only if** they are in a `dict` so that you can reference and enable the proper plugin when needed.

```python
from accelerate.utils import DeepSpeedPlugin

> zero2_plugin = DeepSpeedPlugin(hf_ds_config="zero2_config.json")

Reviewer 4: Would be good to document somewhere if/how `accelerate launch` works when two configs are needed. As a user, I rely heavily on `accelerate launch --config_file=/path/to/deepspeed_zero{1,2,3}.yaml` and I wonder if we can enable something like:

```
# comma separated?
accelerate launch --config_file=deepspeed_zero2.yaml,deepspeed_zero3.yaml
```

And then the training code assumes the first plugin comes from the first config, etc

Author: We can't do that currently, it is solely based on creating `DeepSpeedPlugin`'s manually.

Author: Most probable, setting up the following:

```yaml
deepspeed_configs:
    config_1:
        deepspeed_config_file: ...
        name: ...
    config_2:
        deepspeed_config_file: ...
        name: ...
```

Reviewer 4: Indeed, having a "master" config file would be nice to have. That way I can have my Z2/Z3 configs fixed and just toggle their use based on the task at hand

```json
{
    "bf16": {
        "enabled": "auto"
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "weight_decay": "auto",
            "torch_adam": true,
            "adam_w_mode": true
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },
    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
    },
    "gradient_accumulation_steps": 1,
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
}
```

<Tip>

    DeepSpeed will raise an error if `train_micro_batch_size_per_gpu` isn't specified, even if this particular model isn't being trained.

</Tip>

From here, create a single [`Accelerator`] and pass in both configurations.

```python
from accelerate import Accelerator

accelerator = Accelerator(deepspeed_plugins=deepspeed_plugins)
```

Now let's see how to use them.

### Student model

> By default, Accelerate sets the first item in the `dict` as the default or enabled plugin (`"student"` plugin). Verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin]` function to see which plugin is enabled.

Reviewer 4: ```suggestion
By default, Accelerate sets the first item in the `dict` as the default or enabled plugin (`"student"` plugin). Verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin`] function to see which plugin is enabled.
```

<!--Copyright 2024 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contains specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.
-->

# Using multiple models with DeepSpeed

<Tip warning={true}>

    This guide assumes that you have read and understand the [DeepSpeed usage guide](./deepspeed.md).

</Tip>

> ## Use cases
> 
> There are a few use cases where one may want to run multiple models when using `accelerate` and `DeepSpeed`:
> 
> * Knowledge distillation
> * RLHF techniques (see the `trl` repo for [more examples](https://github.com/huggingface/trl))
> * Training multiple models at once.
> 
> As it stands, Accelerate has a **very experimental API** to help you accomplish this.
> 
> This tutorial will focus on two common use cases:
> 
> 1. Using `zero2` to train a model, and `zero3` to perform inference for assisting of the training for the `zero2` model (as `zero2` is faster to train on if the model fits onto a single GPU), which is shown in the RLHF example
> 2. Training multiple *disjoint* models at once, shown later

Reviewer 5: ```suggestion
Running multiple models with Accelerate and DeepSpeed is useful for:

* Knowledge distillation
* RLHF techniques (see the [TRL](https://github.com/huggingface/trl) library for more examples)
* Training multiple models at once

Currently, Accelerate has a **very experimental API** to help you use multiple models.

This tutorial will focus on two common use cases:

1. An RLHF use case, using ZeRO2 to train a model and ZeRO3 for inference to assist in training the ZeRO2 model. ZeRO2 is faster to train with if the model fits on a single GPU.
2. Training multiple *disjoint* models at once.
```

> ## Multiple models, only one is trained

Reviewer 5: ```suggestion
## RLHF
```

> RLHF is a good example of such use-case
> 
> Normally, you would use a single [`utils.DeepSpeedPlugin`] at a time, however in this case we have two separate configurations. Accelerate allows you to create and use multiple plugins **if and only if** they are in a `dict` so that we can reference/enable the proper plugin when needed:

Reviewer 5: ```suggestion
RLHF is a good example of using multiple models, but only training one of them.

Normally, you would use a single [`utils.DeepSpeedPlugin`] at a time. However, in this case, there are two separate configurations. Accelerate allows you to create and use multiple plugins **if and only if** they are in a `dict` so that you can reference and enable the proper plugin when needed.
```

```python
from accelerate.utils import DeepSpeedPlugin

zero2_plugin = DeepSpeedPlugin(hf_ds_config="zero2_config.json")
zero3_plugin = DeepSpeedPlugin(hf_ds_config="zero3_config.json")

deepspeed_plugins = {"student": zero2_plugin, "teacher": zero3_plugin}
```

> When doing so,  `zero2_config.json` should be configured for full training (so specifying `scheduler` and `optimizer` if not utilizing your own) while `zero3_config.json` should *only* be configured for the inference model,
> such as the following example for a `zero3_config.json`:

Reviewer 5: ```suggestion
The `zero2_config.json` should be configured for full training (so specify `scheduler` and `optimizer` if you are not utilizing your own), while `zero3_config.json` should only be configured for the inference model, as shown in the example below.
```

```json
{
    "bf16": {
        "enabled": "auto"
    },
    "zero_optimization": {
        "stage": 3,
        "overlap_comm": true,
        "reduce_bucket_size": "auto",
        "stage3_prefetch_bucket_size": "auto",
        "stage3_param_persistence_threshold": "auto",
        "stage3_max_live_parameters": "auto",
        "stage3_max_reuse_distance": "auto",
    },
    "train_micro_batch_size_per_gpu": 1
}
```

> And the zero2 counterpart:

Reviewer 5: ```suggestion
An example `zero2_config.json` configuration is shown below.
```

```json
{
    "bf16": {
        "enabled": "auto"
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "weight_decay": "auto",
            "torch_adam": true,
            "adam_w_mode": true
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },
    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
    },
    "gradient_accumulation_steps": 1,
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
}
```

<Tip>

>     DeepSpeed will raise an error if we don't specify a `train_micro_batch_size_per_gpu`, despite not really 
>     training this particular model

Reviewer 5: ```suggestion
    DeepSpeed will raise an error if `train_micro_batch_size_per_gpu` isn't specified, even if this particular model isn't being trained.
```

</Tip>

> From here, we can create a single [`Accelerator`] and pass in both configurations:

Reviewer 5: ```suggestion
From here, create a single [`Accelerator`] and pass in both configurations.
```

```python
from accelerate import Accelerator

accelerator = Accelerator(deepspeed_plugins=deepspeed_plugins)
```

Now let's see how to use them.

> ### The student model

Reviewer 5: ```suggestion
### Student model
```

> By default, `accelerate` will set the first item in the `dict` as the default/enabled plugin (so here our `"student"` plugin). We can verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin]` function to see which one is being used:

Reviewer 5: ```suggestion
By default, Accelerate sets the first item in the `dict` as the default or enabled plugin (`"student"` plugin). Verify this by using the [`utils.deepspeed.get_active_deepspeed_plugin]` function to see which plugin is enabled.
```

```python
active_plugin = get_active_deepspeed_plugin(accelerator.state)
assert active_plugin is deepspeed_plugins["student"]
```

> Also the `AcceleratorState` will keep the active DeepSpeed plugin saved in `state.deepspeed_plugin`:

Reviewer 5: ```suggestion
[`AcceleratorState`] also keeps the active DeepSpeed plugin saved in `state.deepspeed_plugin`.
```

```python
assert active_plugin is accelerator.deepspeed_plugin
```

> Since the `student` is the currently active plugin, let's go ahead and prepare our model, optimizer, and scheduler like normal:

Reviewer 5: ```suggestion
Since `student` is the currently active plugin, let's go ahead and prepare the model, optimizer, and scheduler.
```

```python
student_model, optimizer, scheduler = ...
student_model, optimizer, scheduler, train_dataloader = accelerator.prepare(student_model, optimizer, scheduler, train_dataloader)
```

> From here, we need to deal with the teacher model.
> 
> ### The teacher model
> 
> First we need to tell the `Accelerator` that the zero3/teacher ZeRO configuration should be used:

Reviewer 5: ```suggestion
Now it's time to deal with the teacher model.

### Teacher model

First, you need to specify in [`Accelerator`] that the `zero3_config.json` configuration should be used.
```

```python
accelerator.state.select_deepspeed_plugin("teacher")
```

> Doing so will disable the `"student"` plugin and enable the `"teacher"` one instead. This will update the
> DeepSpeed stateful config inside of `transformers`, and change which plugin configuration gets called when using
> `deepspeed.initialize()`, allowing us to use the no-code `deepspeed.zero.Init`  context manager `transformers` provides:

Reviewer 5: ```suggestion
This disables the `"student"` plugin and enables the `"teacher"` plugin instead. The
DeepSpeed stateful config inside of Transformers is updated, and it changes which plugin configuration gets called when using
`deepspeed.initialize()`. This allows you to use the no-code `deepspeed.zero.Init`  context manager Transformers provides.
```

```python
teacher_model = AutoModel.from_pretrained(...)
teacher_model = accelerator.prepare(teacher_model)
```

> Otherwise you should manually initialize them under `deepspeed.zero.Init`:

Reviewer 5: ```suggestion
Otherwise, you should manually initialize the model with `deepspeed.zero.Init`.
```

```python
with deepspeed.zero.Init(accelerator.deepspeed_plugin.config):
    model = MyModel(...)
```

### Training

> From here, your training loop can be whatever you like, making sure that `teacher_model` is never being trained on:

Reviewer 5: ```suggestion
From here, your training loop can be whatever you like, as long as `teacher_model` is never being trained on.
```

```python
teacher_model.eval()
student_model.train()
for batch in train_dataloader:
    with torch.no_grad():
        output_teacher = teacher_model(**batch)
    output_student = student_model(**batch)
    # Combine the losses or modify it in some way
    loss = output_teacher.loss + output_student.loss
    accelerator.backward(loss)
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
```

> ## Multiple models all being trained

Reviewer 5: ```suggestion
## Train multiple disjoint models
```

> Multiple models being trained is a more complicated scenario and is actively being looked upon by the team.
> In its current state, we assume that each model is **completely disjoint** from the other during training.
> 
> With this situation, we still require two [`utils.DeepSpeedPlugin`]'s to be made, however we also further
> require a second `Accelerator`, since we need to call different `deepspeed` engines at different times, and a single `Accelerator` only
> carries one instance of it at a time.
> 
> Since the [`state.AcceleratorState`] is a stateful object though, it is already aware of both `DeepSpeedPlugin`'s available,
> meaning you can just instantiate a second `Accelerator` with no extra arguments:

Reviewer 5: ```suggestion
Training multiple models is a more complicated scenario.
In its current state, we assume each model is **completely disjointed** from the other during training.

This scenario still requires two [`utils.DeepSpeedPlugin`]'s to be made. However, you also need a second [`Accelerator`], since different `deepspeed` engines are being called at different times. A single [`Accelerator`] can only carry one instance at a time.

Since the [`state.AcceleratorState`] is a stateful object though, it is already aware of both [`utils.DeepSpeedPlugin`]'s available. You can just instantiate a second [`Accelerator`] with no extra arguments.
```

```python
first_accelerator = Accelerator(deepspeed_plugins=deepspeed_plugins)
second_accelerator = Accelerator()
```

> Similar to before, we can call either `first_accelerator.state.select_deepspeed_plugin()` to enable/disable
> a particular plugin, and call `.prepare` like normal:

Reviewer 5: ```suggestion
You can call either `first_accelerator.state.select_deepspeed_plugin()` to enable or disable
a particular plugin, and then call [`prepare`].
```

# For this example, `get_training_items` is a nonexistent function that gets the setup we need for training
first_optimizer, first_scheduler, train_dl, eval_dl = get_training_items(model1)
first_model, first_optimizer, first_scheduler, train_dl, eval_dl = accelerator.prepare(
    first_model, first_optimizer, first_scheduler, train_dl, eval_dl
)

second_accelerator.state.select_deepspeed_plugin("second_model")
second_model = AutoModel.from_pretrained(...)
# For this example, `get_training_items` is a nonexistent function that gets the setup we need for training
second_optimizer, second_scheduler, _, _ = get_training_items(model2)
second_model, second_optimizer, second_scheduler = accelerator.prepare(
    second_model, second_optimizer, second_scheduler
)
```

And now you can train:

```python
for batch in dl:
    outputs1 = first_model(**batch)
    first_accelerator.backward(outputs1.loss)
    first_optimizer.step()
    first_scheduler.step()
    first_optimizer.zero_grad()
    
    outputs2 = model2(**batch)
    second_accelerator.backward(outputs2.loss)
    second_optimizer.step()
    second_scheduler.step()
    second_optimizer.zero_grad()
```

> ## More Resources

Reviewer 5: ```suggestion
## Resources
```

> To see more examples, please check out the [related tests](https://github.com/huggingface/accelerate/blob/main/src/accelerate/test_utils/scripts/external_deps/test_ds_multiple_model.py) we currently have in `Accelerate`, and we will be improving this further
> with better examples, documentation, and functionality as we continue to flesh out this API

Reviewer 5: ```suggestion
To see more examples, please check out the [related tests](https://github.com/huggingface/accelerate/blob/main/src/accelerate/test_utils/scripts/external_deps/test_ds_multiple_model.py) currently in [Accelerate].
```
